Create custom TextInput

kakajann opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!-- Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing -->
Fixes #3108 

### Tests
<!--- 
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image 
--->
Make sure form inputs match specs & screenshots below

### QA Steps
<!--- 
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image 
--->
Same as above

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Specific Specs

- [x] Text size should be 13pt for field label and 15pt for text
- [x] Keeping our existing border style (blue when focused/active, red when error/active) and button style (green for primary actions)
- [x] Field inside the box should be white (both in a selected field and unselected field)
- [x] Field label text should be grey (both in a selected field and unselected field)
- [x] These field changes need to be made to all fields on the mobile app
- [x] Input/select height should be 52
- [x] The horizontal padding should be 12
- [x] The vertical padding should be 8, though the label/text should be vertically centered

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
https://user-images.githubusercontent.com/11542415/124325468-19da8880-db9e-11eb-96ba-8ab5b569bac0.mov

#### Mobile Web
https://user-images.githubusercontent.com/11542415/124325382-ea2b8080-db9d-11eb-9414-e1d3ee7a4b18.mov

#### Desktop
https://user-images.githubusercontent.com/11542415/124325277-d2ec9300-db9d-11eb-9172-df975a36c98d.mov

#### iOS
https://user-images.githubusercontent.com/11542415/124325142-9c167d00-db9d-11eb-848c-f4d2dc409674.mov

kakajann committed:
create custom textinput

MariaHCD commented:
Un-assigning myself as this isn't ready for review yet

kakajann committed:
merge from main

kakajann committed:
restyling

kakajann committed:
change password screen implementation and small refactors

kakajann committed:
login screen implementation

kakajann committed:
payment page implementation

kakajann committed:
merge from main

kakajann committed:
update all textinputs to expensitextinput

kakajann committed:
resolve conflicts

kakajann committed:
make ExpensiTextInput cross-platform

kakajann committed:
picker styles and implementations

kakajann committed:
picker proptypes refactoring

kakajann committed:
expensiTextInput desktop hotfixes

kakajann committed:
resolve conflicts

kakajann committed:
fix autoFocus issue

kakajann commented:
@shawnborton Could you take a look at this PR?

shawnborton commented:
Can do, will take a pass at this today!

shawnborton commented:
Some weird behavior with Chrome and autofill: 
![image](https://user-images.githubusercontent.com/2319350/124515722-fcbbda80-dd94-11eb-987d-e3ff3578e912.png)


Not entirely sure how to fix or diagnose that.

shawnborton commented:
For select/picker menus, can we make it so that you can tap anywhere on the picker to launch the menu? Currently it looks like it only works if you tap on the value text, but if you tap the label or anything else, it doesn't launch.

shawnborton commented:
Are we planning to do any error highlighting with these inputs? I can see an argument where that's a separate project as I don't believe we currently do that anywhere. 

![image](https://user-images.githubusercontent.com/2319350/124516435-7b654780-dd96-11eb-8f68-a4401183101b.png)


![image](https://user-images.githubusercontent.com/2319350/124516443-7f916500-dd96-11eb-8ecc-d1d93ab365be.png)

shawnborton commented:
For disabled pickers, we should make sure the mouse cursor is not a hand. 
![image](https://user-images.githubusercontent.com/2319350/124516631-fd557080-dd96-11eb-9d8d-c73ec11a0b66.png)

shawnborton commented:
Looks like some space below the "Add Phone Number" and the hint text below it was removed? Or perhaps that happened in a different PR?

Before | After
----- | -----
![image](https://user-images.githubusercontent.com/2319350/124516672-10684080-dd97-11eb-9e99-1a839b728077.png) | ![image](https://user-images.githubusercontent.com/2319350/124516709-24ac3d80-dd97-11eb-98c0-bb33b25dfbd8.png)

shawnborton commented:
@kakajann just a few comments I left above, otherwise this is looking great - nice work!

kakajann committed:
Merge branch 'main' of github.com:studio-504/Expensify.cash into form-styles

kakajann committed:
picker pressable issue

kakajann commented:
> Some weird behavior with Chrome and autofill

I've tried `autoCompleteType="off"` but it didn't work and I also don't know how to fix.

> For select/picker menus, can we make it so that you can tap anywhere on the picker to launch the menu? Currently it looks like it only works if you tap on the value text, but if you tap the label or anything else, it doesn't launch.

Fixed.

> For disabled pickers, we should make sure the mouse cursor is not a hand.

Fixed.

> Looks like some space below the "Add Phone Number" and the hint text below it was removed? Or perhaps that happened in a different PR?

I didn't touch it, maybe it comes from another PR.

>Are we planning to do any error highlighting with these inputs? I can see an argument where that's a separate project as I don't believe we currently do that anywhere.

I don't know the plan) but I implemented a style for errors. simply use `error={true}` to change the border color to red

kakajann committed:
merge from main to resolve conflicts

kakajann committed:
fix lint issue

kakajann commented:
@expensifyBot

shawnborton commented:
To assign PullerBear, you just choose that name from the Reviewers section and then it will automatically pick an Expensify team member:
![image](https://user-images.githubusercontent.com/2319350/125322314-7d795a00-e2f2-11eb-9b6f-bcaa0ce0f6f3.png)

shawnborton commented:
> I don't know the plan) but I implemented a style for errors. simply use error={true} to change the border color to red

This sounds perfect, we can make a follow up issue to implement the logic to use this. Thanks!

> I've tried autoCompleteType="off" but it didn't work and I also don't know how to fix.

I'm also stumped, perhaps @Beamanator has some ideas?

Beamanator commented:
> I've tried autoCompleteType="off" but it didn't work and I also don't know how to fix.

Hmm from my understanding, after reading through lots of SO discussions on [this page](https://stackoverflow.com/questions/11708092/detecting-browser-autofill), it may not be possible / may not be worth taking lots of time figuring out how to make it work on all browsers (since it seems many browsers handle auto-fill differently).

What if we have the "Password" label always floating up on this page, since 999 times out of 1000 users will spend 3 seconds on this page and move right along? Also, the Password textbox gets auto-focused so I don't think users will mind if the "Password" label stays floating up if they click away from the textbox.

What are y'all's thoughts? :D

shawnborton commented:
> What if we have the "Password" label always floating up on this page, since 999 times out of 1000 users will spend 3 seconds on this page and move right along?

I see your point but I would prefer if we don't make special exceptions for our inputs - it would be great if they just had a fairly universal behavior. 

I know it might be a stretch to make sure every browser works the same, but is it possible to at least solve for Chrome/webkit?

Beamanator commented:
I mean [this solution](https://stackoverflow.com/a/24784274/2430414) from SO kinda looks the best - it maaay work cross-browser, I haven't tested. 

It could also be fair to fix this final remaining issue later since I believe a decent amount of other PRs want this update 🤷

shawnborton commented:
> It could also be fair to fix this final remaining issue later since I believe a decent amount of other PRs want this update 🤷

That's a fair point, I don't feel strongly though so happy to do whatever you think is best.

Beamanator commented:
Alright I got a bit of feedback [here](https://expensify.slack.com/archives/C03TQ48KC/p1626297658173200) and I'm proposing we move forward without worrying about the Password field issue, and make an issue to fix it once this PR is merged 👍

kakajann committed:
resolve conflicts

kakajann commented:
@Beamanator merged from `main`

Beamanator reviewed (changes requested):
Most of the form fields look fantastic! Here's 2 that I believe need to be edited (see how the label & text being typed don't align like all other new fields) - Note: This is on iOS

<img width="466" alt="Screen Shot 2021-07-19 at 7 01 51 PM" src="https://user-images.githubusercontent.com/3885503/126198834-9cb86dd2-5f24-4658-a2e2-fcad3bac97bc.png">
<img width="466" alt="Screen Shot 2021-07-19 at 7 01 41 PM" src="https://user-images.githubusercontent.com/3885503/126198846-4876818f-3bc7-4913-ab83-8008a2f01d24.png">

kakajann committed:
add custom translate x prop to expensiInput for label transformation

kakajann commented:
Fixed it

kakajann commented:
I have read the CLA Document and I hereby sign the CLA

kakajann committed:
Trigger Build

kakajann committed:
Merge branch 'main' into form-styles

kakajann committed:
resolve conflicts for the 9th time

kakajann committed:
resolve conflicts for the 10th time

kakajann commented:
Can anybody review this? It's kind of annoying to keep merging from `main`

shawnborton commented:
@Beamanator are you able to give this a final review?

Beamanator commented:
Yeah sorry fellas, I was OOO a decent amount last week - I'll review now 👍 (sorry for the delay @kakajann - can you fix this last merge conflict? hoping it's the last I'll ask you to fix 🙏 )

kakajann commented:
Sure

kakajann committed:
resolve conflicts

kakajann commented:
Done

Beamanator commented:
@kakajann why does it look like you didn't merge the podfile correctly? I see lots of changes in this PR now

kakajann commented:
> @kakajann why does it look like you didn't merge the podfile correctly? I see lots of changes in this PR now

Got this error from Pod:
```
ERROR: Parsing unable to continue due to merge conflicts present in:
the file located at /Users/kakajann/www/Expensify.cash/ios/Podfile.lock
```
and I had to delete the `Podfile.lock` and regenerate

kakajann committed:
revert podfile.lock

Beamanator reviewed (changes requested):
It looks like these password field labels / values are not aligned in android and IOS:

| iOS | Android |
|---|---|
| https://user-images.githubusercontent.com/3885503/127005634-90fc5a5d-3831-47ee-810a-528356eeea13.mov | <img width="397" alt="Screen Shot 2021-07-26 at 4 39 41 PM" src="https://user-images.githubusercontent.com/3885503/127007860-06a6ef53-e292-426e-a380-4128e8a6fb9b.png"> |

Additionally, in Android I'm not seeing the inputs show anything when typing, but I know the text is entering correctly because I'm still able to sign in. Example:

https://user-images.githubusercontent.com/3885503/127008160-c3e35edf-136e-484b-b00a-336589c3aded.mov

I'm checking to see if that last error also exists on `main`

kakajann committed:
android related bugfixes

kakajann commented:
Fixed the issues

Beamanator commented:
Thanks @kakajann - I see there's multiple conflicts again - how about once you fix these conflicts if nothing is super broken I'll merge, then if there's tiny modifications needed I can request those in a new PR? (so the conflicts don't keep happening)

kakajann committed:
merge from main

kakajann committed:
add size prop to ExpensiPicker

kakajann committed:
change ExpensiPicker default prop

kakajann commented:
@Beamanator Merged from main

Beamanator reviewed (changes requested):
@kakajann I can't merge this if there's easily noticeable style issues - did you test your last few commits and notice these issues?

<img width="386" alt="Screen Shot 2021-07-29 at 10 47 11 PM" src="https://user-images.githubusercontent.com/3885503/127563619-e88eb76f-fa92-4321-b927-10adc59769bd.png">
<img width="386" alt="Screen Shot 2021-07-29 at 10 46 18 PM" src="https://user-images.githubusercontent.com/3885503/127563624-7623f8ec-231d-410b-bab3-0c4aa438f718.png">

kakajann commented:
@Beamanator I actually only test it on iOs. I'll check again on every platform now

kakajann committed:
fix desktop related issue

kakajann commented:
@Beamanator fixed the issue

Beamanator commented:
Thanks so much, I'll pull and retest 👍 (yes please test all platforms every time, cuz I do that too haha)

parasharrajat commented on src/styles/styles.js:
```diff
@@ -526,6 +526,7 @@ const styles = {
     },
     expensiTextInputDesktop: {
         outline: 'none',
+        boxShadow: 'none',
```
This will throw up in Native . Please use `addOutlineWidth(0)`

kakajann replied:
Changed

roryabraham commented on src/components/ExpensiTextInput/index.ios.js:
```diff
@@ -0,0 +1,126 @@
+        ]).start();
+    }
+
+    onFocus = () => {
```
Same here

kakajann replied:
Done

roryabraham commented on src/components/ExpensiTextInput/index.ios.js:
```diff
@@ -0,0 +1,126 @@
+        }
+    }
+
+    onBlur = () => {
```
Same here

kakajann replied:
Done

roryabraham commented on src/components/ExpensiTextInput/index.ios.js:
```diff
@@ -0,0 +1,126 @@
+        }
+    }
+
+    focus = () => this.input.focus()
```
I don't think we need this function, but if you _do_ include it make it a normal function instead of an arrow function.

kakajann replied:
Removed

roryabraham commented on src/components/ExpensiTextInput/index.ios.js:
```diff
@@ -0,0 +1,126 @@
+        const {
+            label, value, placeholder, error, containerStyles, androidStyles, ...inputProps
+        } = this.props;
+        const {
```
No need to destructure state here.

kakajann replied:
Removed

roryabraham commented on src/components/ExpensiTextInput/index.ios.js:
```diff
@@ -0,0 +1,126 @@
+            isFocused, labelTranslateY, labelTranslateX, labelScale,
+        } = this.state;
+
+        const hasLabel = !!label.length;
```
```suggestion
        const hasLabel = Boolean(label.length);
```

roryabraham replied:
Just clearer this way in my opinion

kakajann replied:
Changed

roryabraham commented on src/components/ExpensiTextInput/index.ios.js:
```diff
@@ -0,0 +1,126 @@
+                        value={value}
+                        onFocus={this.onFocus}
+                        onBlur={this.onBlur}
+                        placeholder={isFocused || !label ? placeholder : null}
```
This is confusing ... can you include parens to make it a bit clearer?

kakajann replied:
Done

roryabraham commented on src/components/ExpensiTextInput/index.js:
```diff
@@ -0,0 +1,120 @@
+import React, {Component} from 'react';
```
Many of the same comments in the `index.ios.js` implementation are relevant in this one as well.

roryabraham commented on src/components/IOUConfirmationList.js:
```diff
@@ -266,11 +267,11 @@ class IOUConfirmationList extends Component {
                     />
                     <Text style={[styles.p5, styles.textMicroBold, styles.colorHeading]}>
-                        {this.props.translate('iOUConfirmationList.whatsItFor')}
+                        {}
```
🤔 Should we just get rid of this text component entirely?

kakajann replied:
Yes, i missed it

roryabraham commented on src/styles/styles.js:
```diff
@@ -290,18 +290,16 @@ const styles = {
+            paddingVertical: 8,
+            justifyContent: 'center',
+            height: '100%',
+            backgroundColor: '#fff',
```
You should only ever reference `themeColors`, not use hex colors directly.

kakajann replied:
Done

roryabraham commented on src/styles/styles.js:
```diff
@@ -290,18 +290,16 @@ const styles = {
-            color: themeColors.text,
-            height: variables.inputComponentSizeNormal,
-            opacity: 1,
+            paddingTop: 25,
```
Just noting here that prettymuch all of our padding and spacing throughout the app comes in increments of 4. So I feel like this should be `24` instead of `25`, but cc @shawnborton for confirmation.

roryabraham commented on src/styles/styles.js:
```diff
@@ -478,6 +476,62 @@ const styles = {
+    expensiTextInputWrapper: {
+        height: 52,
+    },
+    expensiTextInputContainer: {
```
Looks style is an _exact_ duplicate of `picker.inputIOS`. So we should collude them into a single reusable style.

kakajann replied:
Where can I define a reusable style?

roryabraham replied:
All these styles can be reusable – it's just a question of keeping them generic and giving them generic names. Then they can be composed into more specific styles. This makes it much easier to make major adjustments to the aesthetic in the app in the future, implement multiple themes/modes, etc...

roryabraham commented on src/styles/styles.js:
```diff
@@ -290,18 +290,16 @@ const styles = {
-            opacity: 1,
+            paddingTop: 25,
+            paddingHorizontal: 12,
+            paddingVertical: 8,
```
You've got `paddingTop: 25` _and_ `paddingVertical: 8`, so one is probably overwriting the other. Definitely don't need both.

kakajann replied:
I can change `paddingVertical` to `paddingBottom`

roryabraham replied:
Bump – change this `paddingVertical` to `paddingBottom`

roryabraham commented on src/styles/styles.js:
```diff
@@ -478,6 +476,62 @@ const styles = {
+        height: '100%',
+        backgroundColor: '#fff',
+    },
+    expensiTextInputContainerWithoutLabel: {
```
This can be a much more generic style (in [here](https://github.com/Expensify/App/blob/cfd317fa1b3f965e311fb4723d94ebafd3931b92/src/styles/utilities/spacing.js#L184))

```javascript
pv0: {
    paddingVertical: 0,
}
```

kakajann replied:
Done

roryabraham replied:
Okay, but the style should still be generic. How about this?

```javascript
pv0: {
    paddingTop: 0,
    paddingBottom: 0,
},
```

roryabraham commented on src/styles/styles.js:
```diff
@@ -478,6 +476,62 @@ const styles = {
+        paddingTop: 0,
+        paddingBottom: 0,
+    },
+    expensiTextInputContainerOnFocus: {
```
These should definitely also be much more generic styles to promote style reuse. Something like this:

```javascript
borderColorFocus: {
    borderColor: themeColors.borderFocus,
},
borderColorDanger: {
    borderColor: themeColors.badgeDangerBG,
},
```

kakajann replied:
Done

roryabraham commented on src/styles/styles.js:
```diff
@@ -478,6 +476,62 @@ const styles = {
+            {scale},
+        ],
+    }),
+    expensiTextInput: {
```
I'm 99% sure that there's an existing reusable text style that could be used instead of introducing this one. If there's really not, this style should be named something much more generic.

kakajann replied:
I couldn't find any style to reuse it

roryabraham commented on src/styles/styles.js:
```diff
@@ -494,6 +548,68 @@ const styles = {
+        borderRadius: variables.componentBorderRadiusNormal,
+        borderColor: themeColors.border,
+        justifyContent: 'center',
+        backgroundColor: '#fff',
```
Use `themeColors`

kakajann replied:
Done

roryabraham commented on src/styles/styles.js:
```diff
@@ -494,6 +548,68 @@ const styles = {
+        justifyContent: 'center',
+        backgroundColor: '#fff',
+    },
+    expensiPickerContainerOnFocus: {
```
This style is another exact duplicate of `expensiTextInputContainerOnFocus`. They should both just be something generic and reusable like `borderColorFocus`

kakajann replied:
Done

roryabraham commented on src/styles/styles.js:
```diff
@@ -494,6 +548,68 @@ const styles = {
+            top: 16,
+            right: 12,
+        },
+        inputWeb: {
```
There's definitely some potential to DRY up `inputWeb`, `inputIOS`, and `inputAndroid`, maybe using some existing styles.

kakajann replied:
Done

kakajann committed:
use addOutlineWidth instead of outline: none

kakajann committed:
remove android generated file

kakajann commented on android/app/src/main/java/com/expensify/chat/generated/BasePackageList.java:
```diff
@@ -0,0 +1,13 @@
+package com.expensify.chat.generated;
```
Removed, but I think we should add it to .gitignore

roryabraham replied:
Sure, feel free to add `android/app/src/main/java/com/expensify/chat/generated/*` to the `.gitignore`

kakajann commented on src/components/ExpensiPicker/defaultProps.js:
```diff
@@ -0,0 +1,9 @@
+const expensiPickerDefaultProps = {
```
Done

kakajann committed:
merge defaultProps and propTypes files for ExpensiPicker

kakajann commented on src/components/ExpensiPicker/index.js:
```diff
@@ -0,0 +1,52 @@
+        };
+    }
+
+    toggleFocus = open => this.setState({
```
Done

kakajann committed:
remove toggleFocus on ExpensiPicker and use inline setState on onOpen/Close callbacks

kakajann commented on src/components/ExpensiPicker/index.js:
```diff
@@ -0,0 +1,52 @@
+
+    render() {
+        const {
+            label, value, placeholder, useDisabledStyles, ...pickerProps
```
I can't do that. That would be syntax error. `Unexpected trailing comma after rest element`

kakajann commented on src/components/ExpensiPicker/propTypes.js:
```diff
@@ -0,0 +1,23 @@
+    }),
+
+    /** Picker disabled style */
+    useDisabledStyles: PropTypes.boolean,
```
I renamed it but it's kind of confusing to use 2 props (`disabled`, `isDisabled`)

`disabled` = user can not tap/click
`isDisabled` = to change the style of picker to grey

Yet.. this wasn't my idea. It was like that before this PR.

Do you want me to remove the isDisabled prop or keep it as is?

kakajann commented on src/components/ExpensiTextInput/ExpensiTextInputDefaultProps.js:
```diff
@@ -0,0 +1,12 @@
+const expensiTextInputDefaultProps = {
```
Done

kakajann committed:
more refactors on ExpensiPicker

kakajann commented on src/components/ExpensiTextInput/ExpensiTextInputLabel/ExpensiTextInputLabelPropTypes.js:
```diff
@@ -0,0 +1,17 @@
+import {Animated} from 'react-native';
+
+const expensiTextInputLabelPropTypes = {
+    label: PropTypes.string,
```
Done

kakajann committed:
prop type refactoring for ExpensiTextInput

kakajann commented on src/components/ExpensiTextInput/ExpensiTextInputLabel/index.js:
```diff
@@ -0,0 +1,31 @@
+
+class ExpensiTextInputLabel extends PureComponent {
+    render() {
+        const {
```
Refactored

kakajann commented on src/components/ExpensiTextInput/ExpensiTextInputLabel/index.js:
```diff
@@ -0,0 +1,31 @@
+import styles from '../../../styles/styles';
+import expensiTextInputLabelPropTypes from './ExpensiTextInputLabelPropTypes';
+
+class ExpensiTextInputLabel extends PureComponent {
```
Done

kakajann commented on src/components/ExpensiTextInput/ExpensiTextInputLabel/index.native.js:
```diff
@@ -0,0 +1,30 @@
+import styles from '../../../styles/styles';
+import expensiTextInputLabelPropTypes from './ExpensiTextInputLabelPropTypes';
+
+class ExpensiTextInputLabel extends PureComponent {
```
Done

kakajann committed:
refactor ExpensiTextInputLabel

kakajann committed:
refactor ExpensiTextInputWrapper

kakajann commented on src/components/ExpensiTextInput/ExpensiTextInputWrapper/ExpensiTextInputWrapperPropTypes.js:
```diff
@@ -0,0 +1,23 @@
+import PropTypes from 'prop-types';
```
Done

kakajann commented on src/components/ExpensiTextInput/ExpensiTextInputPropTypes.js:
```diff
@@ -0,0 +1,32 @@
+import PropTypes from 'prop-types';
```
Done

kakajann commented on src/components/ExpensiTextInput/ExpensiTextInputPropTypes.js:
```diff
@@ -0,0 +1,32 @@
+    placeholder: PropTypes.string,
+
+    /** Callback that is called when the text input is focused. */
+    onFocusExtra: PropTypes.func,
```
I kind of need it.

`onFocus` is used in `ExpensiTextInput` component to change border color.
`onFocusExtra` is to do something custom in `SomeOtherComponent`

example usage here: _(src/pages/settings/PasswordPage.js)_
```jsx
<ExpensiTextInput
    label={`${this.props.translate('passwordPage.newPassword')}*`}
    secureTextEntry
    autoCompleteType="password"
    textContentType="password"
    value={this.state.newPassword}
    onChangeText={newPassword => this.setState({newPassword})}
    onFocusExtra={() => this.setState({isPasswordRequirementsVisible: true})}
    onBlurExtra={() => this.setState({isPasswordRequirementsVisible: false})}
/>
```
Do you have suggestion about how to refactor?

kakajann commented on src/components/ExpensiTextInput/ExpensiTextInputPropTypes.js:
```diff
@@ -0,0 +1,32 @@
+    onFocusExtra: PropTypes.func,
+
+    /** Callback that is called when the text input is blurred. */
+    onBlurExtra: PropTypes.func,
```
answered above

kakajann commented on src/components/ExpensiTextInput/ExpensiTextInputPropTypes.js:
```diff
@@ -0,0 +1,32 @@
+    /** Styles for the outermost container for this component. */
+    containerStyles: PropTypes.arrayOf(PropTypes.object),
+
+    /** label translate x */
```
comment comes from here: [TextInputWithFocusStyles](https://github.com/Expensify/App/blob/main/src/components/TextInputWithFocusStyles.js)

containerStyles can be used to customize ExpensiTextInput's wrapper, for  example: add marginTop.

kakajann replied:
I changed the comment now

kakajann committed:
change containerStyles comment

kakajann commented on src/components/ExpensiTextInput/ExpensiTextInputWrapper/ExpensiTextInputWrapperPropTypes.js:
```diff
@@ -0,0 +1,23 @@
+    /** Input label */
+    children: PropTypes.node.isRequired,
+
+    /** Has Label */
```
Done

kakajann commented on src/components/ExpensiTextInput/ExpensiTextInputWrapper/ExpensiTextInputWrapperPropTypes.js:
```diff
@@ -0,0 +1,23 @@
+    /** Has Label */
+    hasLabel: PropTypes.bool.isRequired,
+
+    /** On press to wrapper */
```
Changed

kakajann commented on src/components/ExpensiTextInput/ExpensiTextInputWrapper/index.js:
```diff
@@ -0,0 +1,38 @@
+    containerStyles: [],
+};
+
+class ExpensiTextInputWrapper extends PureComponent {
```
Done

kakajann commented on src/components/ExpensiTextInput/index.android.js:
```diff
@@ -0,0 +1,23 @@
+);
+
+ExpensiTextInputAndroid.propTypes = expensiTextInputPropTypes;
+ExpensiTextInputAndroid.defaultProps = expensiTextInputDefaultProps;
```
Done

kakajann commented on src/components/ExpensiTextInput/index.android.js:
```diff
@@ -0,0 +1,23 @@
+import expensiTextInputPropTypes from './ExpensiTextInputPropTypes';
+import expensiTextInputDefaultProps from './ExpensiTextInputDefaultProps';
+
+const ExpensiTextInputAndroid = ({
```
`Identifier 'ExpensiTextInput' has already been declared `

kakajann committed:
more refactors for ExpensiTextInputWrapper

kakajann commented on src/components/ExpensiPicker/index.js:
```diff
@@ -0,0 +1,52 @@
+        const {
+            label, value, placeholder, useDisabledStyles, ...pickerProps
+        } = this.props;
+        const {isOpen} = this.state;
```
Done

kakajann commented on src/components/ExpensiPicker/propTypes.js:
```diff
@@ -0,0 +1,23 @@
+import PropTypes from 'prop-types';
+
+const propTypes = {
+    label: PropTypes.string,
```
Done

kakajann commented on src/components/ExpensiPicker/propTypes.js:
```diff
@@ -0,0 +1,23 @@
+        label: PropTypes.string,
+    }),
+
+    /** Picker disabled style */
```
Done

kakajann commented on src/components/ExpensiTextInput/index.android.js:
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import styles from '../../styles/styles';
+import ExpensiTextInput from './index.ios';
```
That's good idea. Just implemented it.

kakajann commented on src/components/ExpensiTextInput/index.ios.js:
```diff
@@ -0,0 +1,126 @@
+import React, {Component} from 'react';
```
Refactored

kakajann committed:
refactor ExpensiTextInput

kakajann commented on src/components/ExpensiTextInput/index.ios.js:
```diff
@@ -0,0 +1,126 @@
+import styles from '../../styles/styles';
+import themeColors from '../../styles/themes/default';
+
+const ACTIVE_LABEL_TRANSLATE_Y = -10;
```
I used BaseExpensiTextInput, so constants no longer repeated

kakajann commented on src/components/ExpensiTextInput/index.ios.js:
```diff
@@ -0,0 +1,126 @@
+        this.input = null;
+    }
+
+    animateLabel = (translateY, translateX, scale) => {
```
Done

kakajann committed:
more refactors for ExpensiTextInput

kakajann committed:
use themeColors instead of hex

kakajann committed:
styles refactor

kakajann committed:
refactor ExpensiPicker styles

kakajann committed:
resolve conflicts

Beamanator reviewed (changes requested):
Thanks for working hard to address the requested changes last week. I tested this morning and found the global search & new chat / new group pages are broken, form fields looked good though

![Screen Shot 2021-08-02 at 10 39 42 AM](https://user-images.githubusercontent.com/3885503/127831207-6f7c3cd1-40fb-4e21-98ad-8dc4a0b55482.png)

roryabraham commented on src/styles/styles.js:
```diff
@@ -490,6 +552,47 @@ const styles = {
+        color: colors.gray3,
+    },
+    expensiPickerLabel: {
+        fontSize: variables.fontSizeLabel,
```
Just noticing a pattern here ... the following styles are all prettymuch the same:

- `formHint`
- `popoverMenuDescription`
- `settingsLoginName`

So it looks like we could create a reusable style like this:

```javascript
labelLabelSupporting: {
    fontFamily: fontFamily.GTA,
    fontSize: variables.fontSizeLabel,
    color: themeColors.textSupporting,
},
```

Then eliminate or greatly simplify the following styles:

- `formHint`
- `popoverMenuDescription`
- `settingsLoginName`
- `optionAlternateText`
- `expensiPickerLabel`

kakajann replied:
Done

Beamanator replied:
I like the idea of combining styles, but why the name `labelLabelSupporting`? Does two "Label"s sounds weird or am I reading the variable name wrong?

roryabraham replied:
Hahah IDK what I was thinking there ... probably `textLabelSupporting` to mirror `textLabel`, `textMicro`, `textMicroSupporting`, etc...

kakajann committed:
forward ExpensiTextInput ref for external usage

kakajann commented on src/components/ExpensiPicker/propTypes.js:
```diff
@@ -0,0 +1,34 @@
+import PropTypes from 'prop-types';
```
Done

kakajann commented on src/components/Picker/PickerPropTypes.js:
```diff
@@ -42,7 +42,7 @@ const propTypes = {
 };
 const defaultProps = {
-    useDisabledStyles: false,
+    isDisabled: false,
```
Done

kakajann commented on src/components/ExpensiTextInput/propTypes.js:
```diff
@@ -0,0 +1,47 @@
+    placeholder: PropTypes.string,
+
+    /** Callback that is called when the text input is focused. */
+    onFocusExtra: PropTypes.func,
```
Renamed now

kakajann commented on src/styles/styles.js:
```diff
@@ -16,6 +16,17 @@ import CONST from '../CONST';
+    fontFamily: fontFamily.GTA,
+    fontSize: variables.fontSizeNormal,
+    paddingHorizontal: 12,
+    paddingVertical: 8,
```
It can. Just changed

kakajann commented on src/styles/styles.js:
```diff
@@ -474,6 +483,59 @@ const styles = {
     },
 
+    expensiTextInputWrapper: {
+        height: 52,
```
Done

kakajann commented on src/styles/styles.js:
```diff
@@ -474,6 +483,59 @@ const styles = {
+        paddingTop: 0,
+        paddingBottom: 0,
+    },
+    borderColorFocus: {
```
Done

kakajann commented on src/styles/styles.js:
```diff
@@ -490,6 +552,47 @@ const styles = {
+    expensiPickerContainerOnFocus: {
+        borderColor: themeColors.borderFocus,
+    },
+    expensiPickerContainerDisabled: {
```
Combined

kakajann committed:
more style refactors

Beamanator commented on src/pages/ReimbursementAccount/CompanyStep.js:
```diff
@@ -169,47 +169,47 @@ class CompanyStep extends React.Component {
-                            placeholder={{value: '', label: 'Type'}}
-                        />
+                        <View style={styles.mt4}>
+                            <Picker
```
Is there a reason you kept `Picker` here and didn't use `ExpensiPicker?`

kakajann replied:
I've must been missed it) Implementing now

Beamanator reviewed (dismissed):
Overall tested well on all platforms and code looks quite solid now - leaving for @roryabraham for final approval 👍

roryabraham commented on src/styles/styles.js:
```diff
@@ -511,6 +602,12 @@ const styles = {
 
     noOutline: addOutlineWidth({}, 0),
 
+    labelLabelSupporting: {
```
Sorry, my fault in the last review 🤦 
```suggestion
    textLabelSupporting: {
```

roryabraham commented on src/components/ExpensiTextInput/propTypes.js:
```diff
@@ -0,0 +1,39 @@
+    placeholder: PropTypes.string,
+
+    /** Input with error  */
+    error: PropTypes.bool,
```
Usually it's clearer to prefix a boolean with `isXXX` or `hasXXX` or something to make it obvious it's a boolean.
```suggestion
    hasError: PropTypes.bool,
```

kakajann commented on src/components/ExpensiPicker.js:
```diff
@@ -0,0 +1,78 @@
+                    onOpen={() => this.setState({isOpen: true})}
+                    onClose={() => this.setState({isOpen: false})}
+                    disabled={isDisabled}
+                    // eslint-disable-next-line react/jsx-props-no-spreading
+                    {...pickerProps}
```
I think there's more props than only `size`
`onChange`, `items`, `icon` also passed with `pickerProps`

I should remove `placeholder`, `value` from `this.props` destruction to make the code clearer

kakajann committed:
remove redundant prop destructing from ExpensiPicker

kakajann committed:
change Picker with ExpensiPicker

kakajann committed:
ExpensiTextInput refactors

kakajann committed:
fix a propType typo

kakajann commented:
@roryabraham could you take a look at to latest changes?

roryabraham commented on src/styles/styles.js:
```diff
@@ -490,11 +550,42 @@ const styles = {
+            border: 'none',
+            ...expensiPicker,
+        },
+        inputIOS: {
```
Since these are the same style, do we need both? Can we just make it `inputNative`?

kakajann replied:
I don't think `react-native-picker-select` supports `inputNative`
https://github.com/lawnstarter/react-native-picker-select/blob/ca6488c2eef5c776a8071400c8b7987712d43397/index.d.ts#L36-L40

roryabraham replied:
Okay, but in our code we can provide the same `inputNative` style object [here](https://github.com/Expensify/App/blob/6abfbb93f32e2b4a5e9493ed2beed6e762ac7e17/src/components/Picker/pickerDisabledStyles/index.android.js#L5) and [here](https://github.com/Expensify/App/blob/6abfbb93f32e2b4a5e9493ed2beed6e762ac7e17/src/components/Picker/pickerDisabledStyles/index.ios.js#L5)

roryabraham replied:
As soon as we get this thread resolved we are good to merge 👍

kakajann replied:
I actually don't use `pickerDisabledStyles`.

and just to DRY this:
```js
inputIOS: {
    ...expensiPicker,
},
inputAndroid: {
    ...expensiPicker,
},
```

I'll have to create 4 files. `Picker/pickerStyles/index.js`, `Picker/pickerStyles/index.ios`, `Picker/pickerStyles/index.android.js`, `Picker/pickerStyles/propTypes.js`

_Picker/pickerStyles/index.android.js_ for example:

```js
import {propTypes, defaultProps} from './propTypes';
import styles from '../../../styles/styles';

const pickerStyles = disabled => ({
    ...styles.expensiPicker(disabled),
    inputAndroid: styles.expensiPicker(disabled).inputNative,
});

pickerStyles.propTypes = propTypes;
pickerStyles.defaultProps = defaultProps;
pickerStyles.displayName = 'pickerStyles';

export default pickerStyles;
```

kakajann replied:
but ok.. I'll just do it

kakajann replied:
I'm curious [is this](https://github.com/Expensify/App/pull/3414/commits/adeddc786b5a91dc7014ebf93c21390670bc4a24) worth it?

roryabraham reviewed (commented):
@kakajann Thanks for working to make all the improvements I requested so far! I haven't really done a ton of testing on this PR, but the code is looking good to me at this point 👍. I will leave it to @Beamanator from here 😄

Beamanator reviewed (dismissed):
Tested again on the devices that were giving us trouble before, and everything seems great! Time to merge :D

Beamanator commented:
@roryabraham back to you (since you requested changes recently) - also thanks so much for the help reviewing this beast of a PR!

kakajann committed:
dry inputIOS and inputAndroid

kakajann commented:
@roryabraham requesting a review

roryabraham commented on src/components/Picker/pickerStyles/index.android.js:
```diff
@@ -0,0 +1,13 @@
+    inputAndroid: styles.expensiPicker(disabled).inputNative,
+});
+
+pickerStyles.propTypes = propTypes;
```
This is a style, not a react component. So it doesn't make sense to include `propTypes`, `defaultProps`, or `displayName`.

roryabraham reviewed (changes requested):
I think we took this change a bit too far. Can we undo the last commit, and then just make the simple change of getting rid of `styles.picker.inputAndroid` and `styles.picker.inputIOS`, and replacing them with `styles.picker.inputNative`.

kakajann committed:
remove propTypes from style component

kakajann committed:
resolve conflict

roryabraham commented:
Approved, we can merge as soon as iOS E2E tests pass.

kakajann commented:
> Approved, we can merge as soon as iOS E2E tests pass.

Thank you. I was waiting for e2e tests to request a review again)

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1120800158) to staging by @roryabraham in version: 1.0.83-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|failure ❌
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

Jag96 commented:
Left a note on the issue linked above but also tagging here, this PR is causing a deploy blocker: https://github.com/Expensify/App/issues/4597#issuecomment-897830892. @kakajann please have a look

botify commented:
This has been deployed to production and is now subject to a 7-day regression period.
If no regressions arise, payment will be issued on 2021-08-24.  :confetti_ball:

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1140313603) to production by @isagoico in version: 1.0.85-9 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|failure ❌ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

mananjadhav commented:
I am tagging this PR to highlight an issue fixed [here](https://github.com/Expensify/App/issues/16683). All conditions in ternary expressions or left-hand operands on conditional renders, should be boolean. This PR is one of the PRs that uses conditional render with string operands, hence I am tagging it here for the contributors to check. 

We've also updated the item in the checklist with this [PR](https://github.com/Expensify/App/pull/18069) to avoid this issue in the future.
